GPII-3531: Initial implementation of new CouchDB test harness.

the-t-in-rtf opened the pull request:
See [GPII-3531](https://issues.gpii.net/browse/GPII-3531) for details.

cindyli commented on tests/js/basic-tests.js:
```diff
@@ -0,0 +1,284 @@
+                        },
+                        // confirm that the record exists now and delete the latest revision.
+                        {
+                            listener: "gpii.tests.couchdb.basic.checkRecordAndStartDelete",
```
Is it possible to handle preDelete (the fetch of _rev) internally and silently?

the-t-in-rtf replied:
This behaviour is actually a part of CouchDB, just like you're required to send a revision to update an entry.   express-pouchdb is not entirely faithful to the way in which CouchDB works, which is part of the reason I created this package.  We should be able to confidently know that our code that performs DELETEs will work against a real CouchDB instance.

We shouldn't make the harness handle DELETEs any differently than CouchDB does, but I understand that this double-request process can be cumbersome.  I haven't found anyone actually performing real deletes in universal or my own work, but once someone needs to do so,  we should create a variation on the CouchDB dataSource that can either make use of the DELETE method, or that can pass the special `_deleted` field as part of an update request.

cindyli commented on tests/js/basic-tests.js:
```diff
@@ -0,0 +1,284 @@
+                    sequence: [
+                        // The record should not exist before we create it.
+                        {
+                            func: "{preInsertRequest}.send",
```
Same as the last comment: is it possible to handle preInsert internally?

the-t-in-rtf replied:
This check is making sure that the data doesn't already exist, i.e. that nothing is left over from a previous run.  Actual inserts by real users will not require an initial lookup.

cindyli commented on src/js/isCouchReady.js:
```diff
@@ -0,0 +1,62 @@
+                    couchReadyPromise.resolve();
+                }
+            },
+            couchReadyPromise.reject
```
It would be helpful to provide an error detail when a failure occurs. The same comment applies to other promise rejections or failure handling.

the-t-in-rtf replied:
This setup passes through the details from the actual rejection.  All of the direct calls of reject do provide error details.

cindyli commented on tests/js/basic-tests.js:
```diff
@@ -0,0 +1,284 @@
+    deleteRequest.send({}, { termMap: { rev: record._rev } });
+};
+
+fluid.defaults("gpii.tests.couchdb.basic.caseHolder", {
```
Would be nice to have tests for view operations for `gpii.test.couchdb.caseHolder` too.

the-t-in-rtf replied:
Good point, and done.

the-t-in-rtf commented:
@cindyli and @sgithens, this pull now includes a feature you've asked about, i.e. support for both the "bulk docs" format (all records wrapped in a "docs" element) and support for the format used with payloads in universal, where the file is simply an array of records.

amb26 commented:
Even after setting the environment variable in Windows, I get numerous failures, of which the first is:
````
16:53:07.936:  Failure in task fixture {
    "task": "{harness}.startup",
    "resolve": "fluid.log",
    "resolveArgs": [
        "Harness startup successful."
....
}: promise has received disposition "reject" with payload Cannot continue, an instance of CouchDB not controlled by our worker is listening
on our port. when "resolve" was expected
16:53:07.998:  jq: FAIL: Module "Testing docker test harness." Test name "Testing loading CouchDB root." - Message: Failure in task fixture
````
The other failures are similar. Although the test run finishes with a coverage report, it does not report a count of passed and failed tests.

the-t-in-rtf commented:
Thanks, @amb26, I will attempt to reproduce in a Windows VM shortly.

the-t-in-rtf commented:
We discussed the problems in depth via chat, turns out the line-parsing was not capable of handling CRLFs.  This has been addressed, and a further timing problem I discovered in testing with Vagrant has also been fixed.  @amb26, this should be ready for further testing.

the-t-in-rtf commented on src/js/harness.js:
```diff
@@ -292,25 +292,34 @@ gpii.test.couchdb.harness.clearInterval = function (that) {
     }
 };
 
+gpii.test.couchdb.harness.monitorContainerOnce = function (that) {
```
In discussions with @amb26, we discussed improving this, his (lightly edited) remarks below:

> The logic of this set of functions, collectively, is pretty obscure.  I recommend making a "timer" subcomponent that embodies the timer, which represents clearing the timer by its own destruction.  There are a few patterns for this in  gpii-app:
> 
> https://github.com/GPII/gpii-app/blob/master/src/shared/utils.js#L35

amb26 commented on src/js/harness.js:
```diff
@@ -20,15 +20,21 @@ fluid.registerNamespace("gpii.test.couchdb.harness");
-        }
-    }, startPromise.reject);
+    // If our worker has already been destroyed, startup is not needed.
+    if (that.worker.lifecycleStatus === "destroyed") {
```
A cleaner check is fluid.isDestroyed(that.worker) - the lifecycleStatus flag does not have API status

the-t-in-rtf replied:
Yup, fixed.

the-t-in-rtf commented:
@amb26, I discovered that this work would not work with "potentia ii", and updated it accordingly.  The change is not complex, just a few updated IoC references to break a circular reference. See [GPII-3945](https://issues.gpii.net/browse/GPII-3945) for details.

the-t-in-rtf commented:
Anyone using this work with "potentia ii" will need this dev release or newer:

1.0.0-dev.20190524T090918Z.5cd04e7.GPII-3531

the-t-in-rtf commented:
This pull now includes a temporary fix for [GPII-3989](https://issues.gpii.net/browse/GPII-3989).  I cut a dev release (`1.0.0-dev.20190627T080736Z.0e3a1fd.GPII-3531`) and updated [the associated universal validation performance pull](https://github.com/GPII/universal/pull/785) to use the new dev release.

the-t-in-rtf commented:
@amb26, I will shortly need to resume work on this project (see [GPII-4083](https://issues.gpii.net/browse/GPII-4083)), we'll either need to finish this one off or agree that further work will be added to the existing review.

amb26 commented on src/js/harness.js:
```diff
@@ -399,61 +320,66 @@ fluid.defaults("gpii.test.couchdb.harness", {
             priority: "last",
             func:     "{that}.events.onDbProvisioningComplete.fire"
         },
+        //"combinedStartup.logHarnessStart": {
```
Stray log comment

the-t-in-rtf replied:
Removed, throughout.

amb26 commented on src/js/harness.js:
```diff
@@ -399,61 +320,66 @@ fluid.defaults("gpii.test.couchdb.harness", {
-            funcName: "gpii.test.couchdb.harness.startMonitoring",
-            args:     ["{that}"]
-        },
+        //"combinedStartup.logProvisioningComplete": {
```
Several more

amb26 commented:
I tried "vagrant ci test linux" several times and sometimes got races like 
````
==> linux: 21:06:23.603:  jq: FAIL: Module "Testing docker test harness." Test name "Testing loading CouchDB root." - Me
ssage: Failure in task fixture {
==> linux:   "task": "{harness}.startup",
==> linux:   "resolve": "fluid.log",
==> linux:   "resolveArgs": [
==> linux:     "Harness startup successful."
==> linux:   ],
==> linux:   "contextThat": {
==> linux:     "typeName": "gpii.test.couchdb.sequenceElement.startHarness",
==> linux: }: promise has received disposition "reject" with payload Timed out waiting for URL 'undefined' to respond. w
hen "resolve" was expected
==> linux: 21:06:23.603:  jq: Source:     at pok (/vagrant/node_modules/infusion/tests/test-core/jqUnit/js/jqUnit.js:112
````

but on each run was unable to get past the "search" test which always times out as follows:
````
==> linux: 21:08:46.198:  jq: Test concluded - Module "Testing view handling." Test name "Reduce only records matching a
 key.": 3/3 passed - PASS
==> linux: 21:08:47.774:  Running using 'docker' worker.
==> linux: 21:08:51.206:  Test case listener has not responded after 5000ms - at sequence pos 1 of 4 sequence element {
==> linux:     "task": "{harness}.startup",
==> linux:     "resolve": "fluid.log",
==> linux:     "resolveArgs": [
==> linux:         "Harness startup successful."
==> linux:     ],
==> linux:     "contextThat": {
==> linux:         "typeName": "gpii.test.couchdb.sequenceElement.startHarness",
....
==> linux:     }
==> linux: } of fixture A search that matches data should return results.
````

amb26 commented:
I got better results running on the real Windows machine via GPII_TEST_COUCH_USE_EXTERNAL - only one test failed:
````
22:17:10.575:  jq: FAIL: Module "Failure mode tests for promise-returning functions." Test name "Test low-level error ha
ndling in `gpii.test.couchdb.runCommandAsPromise`." - Message: The promise should not have been resolved.
22:17:10.579:  jq: Source:     at pok (E:\Source\gits\gpii-couchdb-test-harness\node_modules\infusion\tests\test-core\jq
Unit\js\jqUnit.js:112:15)
    at Object.fail (E:\Source\gits\gpii-couchdb-test-harness\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:1
29:13)
    at Array.<anonymous> (E:\Source\gits\gpii-couchdb-test-harness\tests\js\failure-mode-tests.js:26:24)
    at Object.that.complete (E:\Source\gits\gpii-couchdb-test-harness\node_modules\infusion\src\framework\core\js\FluidP
romises.js:72:25)
    at Object.that.resolve (E:\Source\gits\gpii-couchdb-test-harness\node_modules\infusion\src\framework\core\js\FluidPr
omises.js:54:22)
    at E:\Source\gits\gpii-couchdb-test-harness\src\js\runCommand.js:27:140
    at ChildProcess.exithandler (child_process.js:195:7)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:886:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
22:17:10.599:  jq: Test concluded - Module "Failure mode tests for promise-returning functions." Test name "Test low-level error handling in `gpii.test.couchdb.runCommandAsPromise`.": 0/1 passed - FAIL
````

However the coverage stats seemed rather poor:

````
=============================== Coverage summary ===============================
Statements   : 81.27% ( 256/315 )
Branches     : 57.3% ( 51/89 )
Functions    : 70.59% ( 36/51 )
Lines        : 81.41% ( 254/312 )
================================================================================
````

the-t-in-rtf commented:
Based on recent comments from @amb26, I updated the README to better reflect the various ways to run the tests in the package.  I also found a bug when running the tests with the `GPII_TEST_COUCH_USE_VAGRANT` environment variable set, which I fixed.

the-t-in-rtf commented:
Just to respond to your comments about coverage, here are the results when running using `vagrant ci test linux`:

```
==> linux: =============================== Coverage summary ===============================
==> linux: Statements   : 89.46% ( 280/313 )
==> linux: Branches     : 66.29% ( 59/89 )
==> linux: Functions    : 82.35% ( 42/51 )
==> linux: Lines        : 89.74% ( 280/312 )
==> linux: ================================================================================
```

These are higher (but not 100%) because they exercise the docker-specific infrastructure, which has slightly more lines of unique code.  Your setup will inevitably report lower coverage because it only uses (another) one of the three infrastructures.  Unless we somehow run all three and combine the results, we're not going to get complete coverage.

amb26 commented:
ok to test

amb26 commented:
CI has failed in a rather strange way: 
````
npm ERR! path /vagrant/node_modules/nyc/node_modules/yargs/node_modules/find-up/package.json.610819665
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open
npm ERR! enoent ENOENT: no such file or directory, open '/vagrant/node_modules/nyc/node_modules/yargs/node_modules/find-up/package.json.610819665'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 
````

amb26 commented:
ok to test

the-t-in-rtf commented:
The tests passed but the vm deletion failed.  I may need to prepare a pull to update the CI config to specify that the `linux` VM should be halted/destroyed.  Ok to test again, let's see if it was a fluke.

the-t-in-rtf commented:
Although it may not matter, it seemed most correct to go ahead and put in [a pull to specify which VM should be halted/destroyed](https://github.com/GPII/ci-service/pull/67).

the-t-in-rtf commented:
@amb26, looks like the delete stage will fail until you merge [the other pull](https://github.com/GPII/ci-service/pull/67).

the-t-in-rtf commented:
Antranig just merged the CI pull that should fix the "delete" stage. OK to test.

amb26 commented:
I guess it's ok to push this in now? "Perfection is achieved on the point of collapse"

the-t-in-rtf commented:
@amb26, we didn't talk about it, but my feeling was that we were holding off to see how the longer term stability in CI, etc. was, as there were non-trivial issues in the beginning. I think it's been long enough that we can merge and cut at least the initial release.
